### PR TITLE
ReRender Optimizations & removal of React Context

### PR DESCRIPTION
**Pull request by DroopyTersen**
With the changes in this, we can now control Components so that they will only re-render if the slice of context/store they care about changes.

Another change was moving away from React Context in favor of a vanilla JS `Observable` object so we can easily subscribe to context changes, then check to see if our selectors return a new result, and if they do, re-render the component that used that selector.

There are 2 new hooks you can use

*useLaunchService*
``` javascript
let calculator = useLaunchService("calculator");
// gives you { value, increment, decrement }
```

There is also a more granular `useLaunchSelector`
``` javascript
let value = useLaunchSelector((ctx) => ctx.state.calculator.value);
// the component using this hook will only re-render if calculator value changes
```

I put the calculator display component back in too to make sure pure "presentation" components work as expected as well.

In the logged output you can see that when I updated the Name, only the NameForm re-rendered. Then when I updated the counter, only the Calculator Components rerendered
image

### CODE REVIEW

**Comment by DroopyTersen** [comment]
Don't feel like you have to merge this. I was more just trying to marry the idea's i've been having about Observables with the API discussions we've been having, with the "Single Context" approach (even though it's no longer a React Context)

File: example/src/App.tsx [hunk 1]
```diff
@@ -1,22 +1,20 @@
...
+import { useLaunchProvider } from "launch.io";
 import Main from "./views/Main";
 import nameService from "./features/name/name.service";
 import calculatorService from "./features/calculator/calculator.service";
 import "./App.css";
 
+const services = [nameService, calculatorService];
+const options = { enableTimeTravel: true };
 export default function App() {
+  useLaunchProvider(services, options);
```
Comment (DroopyTersen) [inline example/src/App.tsx:11]: Notice, you no longer need a `Provider` component wrapper. you just have to use this hook at the top

File: example/src/features/calculator/CalculatorDisplay.tsx [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import React, { useEffect } from "react";
+import { useLaunchSelector } from "launch.io";
+
+const CalculatorDisplay = () => {
+  const value = useLaunchSelector<number>(
+    (context) => context.state.calculator.value
+  );
```
Comment (DroopyTersen) [inline example/src/features/calculator/CalculatorDisplay.tsx:7]: This is an example of the more granular `useLaunchSelector`.  There is also `useLaunch`, which returns the entire context, and `useLaunchService` which returns a specific service's state and actions merged together.

File: example/src/features/name/NameForm.tsx [hunk 3]
```diff
@@ -1,30 +1,38 @@
-import React from "react";
-import { useName } from "./name.hooks";
+import React, { useEffect } from "react";
+import { useLaunchService } from "launch.io";
 
 const NameForm = () => {
-  const { firstName, lastName, updateName } = useName();
+  const name = useLaunchService("name");
+
```
Comment (DroopyTersen) [inline example/src/features/name/NameForm.tsx:6]: I was playing around with what it felt like to NOT destructure the result

File: example/src/views/Main.tsx [hunk 4]
```diff
@@ -1,14 +1,16 @@
 import React from "react";
 import NameForm from "../features/name/NameForm";
 import CalculatorForm from "../features/calculator/CalculatorForm";
+import CalculatorDisplay from "../features/calculator/CalculatorDisplay";
 import HistoryForm from "../features/history/HistoryForm";
 
-export default function Main() {
+export default React.memo(function Main() {
```
Comment (DroopyTersen) [inline example/src/views/Main.tsx:7]: The top level `App` component that has the `useLaunchProvider` is going to rerender with every context change. So memoizing the direct children can short circuit the rerender

**Comment by DroopyTersen** [comment]
@jbillmann  i just pushed an update that broke something with the history. I have to run, but I'll fix it tomorrow

**Comment by jbillmann** [comment]
@DroopyTersen Awesome!  I have to digest this PR over the next day or two.  After putting in a bunch of time with React.Context, I feel like Launch.IO is/was stuck between a rock and a hard place at times.  It had me shaking my head and pondering alternative options.

**Comment by DroopyTersen** [comment]
@jbillmann i fixed up a couple bugs related to the history service. something is still wonky with replaying the actions. I think it is related to state updates being async, so when you replay the actions its like you call them all concurrently instead of in sequence.

**Comment by DroopyTersen** [comment]
Ahh I figured out the wonkyness. It's when we replay all the actions, we get the updated state, but part of that is triggering the call to fetch the name from the api, so that call gets back after all the actions are replayed, and resets the name.

**Comment by jbillmann** [comment]
@DroopyTersen So, this might be a code smell, but I wanted to keep things simple in spite of that during initial development.  Basically, if a service action does not return state, Launch.IO assumes it launched other actions that did.  Therefore, we only replay actions that we were launched and return state.

**Comment by DroopyTersen** [comment]
> @DroopyTersen So, this might be a code smell, but I wanted to keep things simple in spite of that during initial development. Basically, if a service action does not return state, Launch.IO assumes it launched other actions that did. Therefore, we only replay actions that were launched and return state.

What I was seeing (and it's totally possible I introduced a regression to cause this behavior) was that we were just checking if the action returned a truthy value. And once you mark an actions as `async` it will automatically return a `Promise` which is truthy. I just enhanced to truthy check with a `Promise` check

**Comment by jbillmann** [comment]
> > @DroopyTersen So, this might be a code smell, but I wanted to keep things simple in spite of that during initial development. Basically, if a service action does not return state, Launch.IO assumes it launched other actions that did. Therefore, we only replay actions that were launched and return state.
> 
> 
> 
> What I was seeing (and it's totally possible I introduced a regression to cause this behavior) was that we were checking if the action returned a truthy value (the updated state). But when you mark an action function as `async`, it will automatically return a `Promise` which is truthy too. I enhanced to check for truthy and not a `Promise`

The truthy check (on `updatedState`) needed to be tightened, regardless of your PR, so your changes didn't introduce any regressions. I'm surprised we didn't see this behavior, before.  In either case, I think your change makes sense.  I will add a separate ticket to possibly further refine the check.

**Pull request merged**